src/include/module/module/base.h

```diff
@@ -0,0 +1,142 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright(c) 2020 - 2023 Intel Corporation. All rights reserved.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *	   Adrian Warecki <adrian.warecki@intel.com>
+ */
+
+#ifndef __MODULE_MODULE_GENERIC__
+#define __MODULE_MODULE_GENERIC__
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+
+#include "interface.h"
+#include "../ipc4/base-config.h"
+
+#define module_get_private_data(mod) ((mod)->priv.private)
+#define module_set_private_data(mod, data) ((mod)->priv.private = data)
+
+/**
+ * \struct module_config
+ * \brief Module config container, used for both config types.
+ */
+struct module_config {
+	size_t size; /**< Specifies the size of whole config */
+	bool avail; /**< Marks config as available to use.*/
+	void *data; /**< tlv config, a pointer to memory where config is stored. */
+	const void *init_data; /**< Initial IPC configuration. */
+#if CONFIG_IPC_MAJOR_4
+	struct ipc4_base_module_cfg base_cfg;
+#endif
+};
+
+/** private, runtime module data */
+struct module_data {
+	void *private; /**< self object, memory tables etc here */
+	struct module_config cfg; /**< module configuration data */
+
+	/* Fields below can only be accessed by the SOF */
+#ifdef SOF_MODULE_API_PRIVATE
+	enum module_state state;
+	size_t new_cfg_size; /**< size of new module config data */
+	void *runtime_params;
+	const struct module_interface *ops; /**< module specific operations */
+	struct module_memory memory; /**< memory allocated by module */
+	struct module_processing_data mpd; /**< shared data comp <-> module */
+	void *module_adapter; /**<loadable module interface handle */
+	uint32_t module_entry_point; /**<loadable module entry point address */
+#endif /* SOF_MODULE_PRIVATE */
+};
+
+/* module_adapter private, runtime data - public version*/
+struct processing_module {
+	struct module_data priv; /**< module private data */
+	uint32_t period_bytes; /** pipeline period bytes */
+
+	/* Fields below can only be accessed by the SOF */
+#ifdef SOF_MODULE_API_PRIVATE
```

Comment:
Is this new, wont it change the ABI if its defined for some users and not others ?

Reply:
@lgirdwood This symbol is defined in the file included by sof. Loadable modules should not define it and nor refer to this part of the structure. This structures should be divided in a future so sof can keep their private fileds in a different structure. It will take a lot of work and will require a lot of changes. Would it be enough for now to add a comment that this is a temporary solution until these fields are separated into a new structure?

Reply:
Yeah, the comment need to be very visible and explain the roadmap forwards. That way its really obvious to everyone working on modules.